packages/loader/container-loader/src/container.ts

```diff
@@ -1359,11 +1373,37 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         protocol.quorum.on("addMember", (clientId, details) => {
             // This is the only one that requires the pending client ID
             if (clientId === this.pendingClientId) {
-                this.setConnectionState(ConnectionState.Connected);
+                let startTime: number | undefined;
+                if (this.prevClientLeftP?.isCompleted === false) {
+                    startTime = performance.now();
+                }
+                // Wait for previous client to leave the quorum before firing "connected" event.
+                if (this.prevClientLeftP) {
+                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                    this.prevClientLeftP.promise.then((leaveReceived: boolean) => {
+                        this.logger.sendTelemetryEvent({
+                            eventName: "ConnectedAfterWait",
+                            duration: startTime !== undefined ? performance.now() - startTime : 0,
+                            timeout: !leaveReceived,
+                            outstandingOps: this._deltaManager.clientSequenceNumber -
+                                this._deltaManager.clientSequenceNumberObserved,
+                        });
+                        this.setConnectionState(ConnectionState.Connected);
```

Comment:
Ops can be processed in disconnected state (through storage). Current logic will set connected state to "connected" when this promise is resolved, irrespective of if we have connection or not. This is wrong. Same for timeout.
The reason that ELSE clause us correct is because it's synchronous - we checked clientId === this.pendingClientId above, and I believe (need to double check) that pendingClientId is undefined if we do not have connection.